[ipcamera] Add channel Privacy Mode to Amcrest and Dahua handlers

sidamos opened the pull request:
Signed-off-by: Thomas Börkel <thomas@boerkel.de>

This adds a new channel to the Amcrest/Dahua things for enable/disable of the Privacy Mode of newer Amcrest/Dahua cameras.
If changed by another application, it is automatically reflected in the OH item.

fwolter commented:
@Skinah Can you take a look?

Skinah reviewed (changes requested):
A few comments:

1. case CHANNEL_FFMPEG_MOTION_CONTROL: I removed that last week and it is now back in your PR. It needs to be removed as it is dead code that can not be reached.

2. LeFunctionStatusSync Is that French that wont work if the camera is set to English?

3. else  {
                    ipCameraHandler.setChannelState(CHANNEL_ENABLE_PRIVACY_MODE, OnOffType.OFF);
                }
Can we assume if we dont find true that the feature is off to prevent needing to look for a second string saving CPU time?

4.<label>Privacy Mode ON/OFF</label> change this to <label>Enable Privacy Mode</label> as the channels label is used to create items from and I like to see consistent names :) 

5. Can you implement a http request on a REFRESH command that asks the camera the state and then updates the controls? it is possible people change settings in an app or camera directly so it helps to keep openHABs controls matching.

cpmeister reviewed (approved):
LGTM

Skinah commented on bundles/org.openhab.binding.ipcamera/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -2663,6 +2665,12 @@
 
+	<channel-type id="enablePrivacyMode">
+		<item-type>Switch</item-type>
+		<label>Privacy Mode ON/OFF</label>
```
```suggestion
		<label>Enable Privacy Mode</label>
```

Skinah commented on bundles/org.openhab.binding.ipcamera/src/main/java/org/openhab/binding/ipcamera/internal/AmcrestHandler.java:
```diff
@@ -205,6 +213,28 @@ public void handleCommand(ChannelUID channelUID, Command command) {
                     ipCameraHandler.sendHttpGET("/cgi-bin/configManager.cgi?action=setConfig&AlarmOut[1].Mode=0");
                 }
                 return;
+            case CHANNEL_FFMPEG_MOTION_CONTROL:
+                if (OnOffType.ON.equals(command)) {
+                    ipCameraHandler.motionAlarmEnabled = true;
+                } else if (OnOffType.OFF.equals(command) || DecimalType.ZERO.equals(command)) {
+                    ipCameraHandler.motionAlarmEnabled = false;
+                    ipCameraHandler.noMotionDetected(CHANNEL_MOTION_ALARM);
+                } else {
+                    ipCameraHandler.motionAlarmEnabled = true;
+                    ipCameraHandler.motionThreshold = Double.valueOf(command.toString());
+                    ipCameraHandler.motionThreshold = ipCameraHandler.motionThreshold / 10000;
+                }
+                ipCameraHandler.setupFfmpegFormat(FFmpegFormat.RTSP_ALARMS);
+                return;
```
```suggestion

```

Skinah commented on bundles/org.openhab.binding.ipcamera/src/main/java/org/openhab/binding/ipcamera/internal/DahuaHandler.java:
```diff
@@ -251,6 +259,27 @@ public void handleCommand(ChannelUID channelUID, Command command) {
                     ipCameraHandler.sendHttpGET("/cgi-bin/configManager.cgi?action=setConfig&AlarmOut[1].Mode=0");
                 }
                 return;
+            case CHANNEL_FFMPEG_MOTION_CONTROL:
+                if (OnOffType.ON.equals(command)) {
+                    ipCameraHandler.motionAlarmEnabled = true;
+                } else if (OnOffType.OFF.equals(command) || DecimalType.ZERO.equals(command)) {
+                    ipCameraHandler.motionAlarmEnabled = false;
+                    ipCameraHandler.noMotionDetected(CHANNEL_MOTION_ALARM);
+                } else {
+                    ipCameraHandler.motionAlarmEnabled = true;
+                    ipCameraHandler.motionThreshold = Double.valueOf(command.toString()) / 10000;
+                }
+                ipCameraHandler.setupFfmpegFormat(FFmpegFormat.RTSP_ALARMS);
+                return;
```
```suggestion

```

sidamos commented:
> 1. case CHANNEL_FFMPEG_MOTION_CONTROL: I removed that last week and it is now back in your PR. It needs to be removed as it is dead code that can not be reached.

That is strange. GitHub gave me a merge conflict and so I thought, this was new code. It is removed now by you, thanks.

> 2. LeFunctionStatusSync Is that French that wont work if the camera is set to English?

No, the camera runs in English.

> 3. else  {
>    ipCameraHandler.setChannelState(CHANNEL_ENABLE_PRIVACY_MODE, OnOffType.OFF);
>    }
>    Can we assume if we dont find true that the feature is off to prevent needing to look for a second string saving CPU time?

Yes, but because of the REFRESH implementation, I had to change that anyway. When someone else changes the Privacy Mode, the camera sends 2 different responses to the alarm channel. I am using now the other string, because it is shorter/easier ("Code=LensMaskOpen;"). For checking the answer of the REFRESH command, I have added check for "table.LeLensMask[0].Enable=true".
 
> 4.Privacy Mode ON/OFF change this to Enable Privacy Mode as the channels label is used to create items from and I like to see consistent names :)

I copied from "Alarm Output 1 ON/OFF". ;-)
But it is changed now by you.

> 5. Can you implement a http request on a REFRESH command that asks the camera the state and then updates the controls? it is possible people change settings in an app or camera directly so it helps to keep openHABs controls matching.

Implemented. On that note: Could the binding send a REFRESH to all channels during startup of openHAB?

Skinah commented:
I am happy with all the changes, thanks for contributing this. I have looked it over and don't see any changes for the GET request being sent on the REFRESH command which this is not necessary, just a nice touch to add it whilst someone with a camera can test it works. 

Yes adding the refresh part can be done to any channel and the GET request is only sent if the channel is linked, however there is a downside, and that is the reply has to go through the multiple IF statements looking for a match, so for common channels that everyone uses they should go at the top so they match quickly and any uncommon ones should go towards the last IF. I need to look to see if there is a way to  improve this at some point and is why I have not added everything as it means more work if it gets reworked.

I see your having some issues with the signoff and the extra commits showing up. See here:
http://schacon.github.io/history

You can do a R reword or E edit on the commit messages to add your signoff.
You can do a D drop on the commits that you want to remove.

`git status` is handy to run if you get lost as it will help guide what you need to do next.

Hope that helps as I am still learning git slowly.

sidamos commented:
> Yes adding the refresh part can be done to any channel and the GET request is only sent if the channel is linked, however there is a downside, and that is the reply has to go through the multiple IF statements looking for a match, so for common channels that everyone uses they should go at the top so they match quickly and any uncommon ones should go towards the last IF. I need to look to see if there is a way to improve this at some point and is why I have not added everything as it means more work if it gets reworked.

OK and could the binding send automatically REFRESH to all channels during startup of openHAB? So that openHAB gets changes that have been made externally while openHAB was down.

Also, it is strange that the camera reports external changes of privacy mode through the alarm stream but it does not report external change of motion detection (on/off). For this, you have to send REFRESH to that channel.

I think, you still need to approve my changes (still learning GitHub pull requests, as you have seen from my commits and force pushes ;-) ).
